Removing networkReady check in StopPodSandbox before calling cni DEL

lyveng opened the pull request:
Race condition is between RunPodSandbox and StopPodSandbox when setting up network.

neerajb reviewed (commented):
Why are we taking a global lock? Can we take a mutex per podSandbox?

lyveng commented:
I realised that there is already a per pod mutex being used inside PluginManager(inside dockershim). Adding one more per pod mutex in docker_sandbox didn't seem like a good idea.

I went through the cni spec. As per the spec, the cni add and del must be ordered such that a ADD always eventually followed by a DEL and there can be multiple DELs as well which need to be handled idempotently by the plugins. Quoting from [cni spec 0.4.0](https://github.com/containernetworking/cni/blob/spec-v0.4.0/SPEC.md#general-considerations):

>The container runtime must order ADD and DEL operations for a container, such that ADD is always eventually followed by a corresponding DEL. DEL may be followed by additional DELs but plugins should handle multiple DELs permissively (i.e. plugin DEL should be idempotent).

In the interest of avoiding multiple calls to cni DEL, PR kubernetes/kubernetes#52864 relied on the networkReady flag to skip cni DEL in StopPodSandbox. Unfortunately, the networkReady flag can be false in two cases:

1. the cni ADD has started executing, but not completed
2. the cni DEL has successfully completed

But the problem with this is that, in some cases like when the StopPodSandbox gets called when cni ADD is currently executing(happens when pod is created and deleted immediately while the cni plugin is latent), the StopPodSandbox skips the cni DEL(point number 1). This is in violation of the CNI spec that I've quoted above. In fact the CNI spec allows the cni DEL being called multiple times and it is the responsibility of the cni plugin to ensure the DEL operation is idempotent.

So I think it would make sense to revert only this part from the original PR kubernetes/kubernetes#52864. I've tested this on my dev cluster and it works fine.

lyveng committed:
Removed networkReady check before calling cni DEL in stop pod sandbox

The networkReady value is false in the following cases:
1. the cni ADD has started executing, but not completed
2. the cni DEL has successfully completed

The check in StopPodSandbox flow only considers reason number 2 and
ignores reason number 1. This check can be removed because as per the
cni spec, it is legal to call cni del multiple times, while cni DEL must
be called at least once after cni ADD is called. When reason number 1
happens, the check was causing cni DEL to be not called at all
effectively leaking ips in ipam.

neerajb commented:
LGTM
